### ISSUE TITLE
Strange output problem

### ISSUE THREAD

**Issue by jrogawa** [JoshOBrien/exiftoolr#24]
Hi there.

I'm having a strange problem that I THINK is related to how exiftoolr is capturing output.

I'm running it from R in this way:
all_tags <- exif_read(path="Files", recursive=T, tags=c("SourceFile", "FileName", "FileModifyDate", "FileCreateDate", "DateTimeOriginal", "CreateDate", "ModifyDate"))

When I'm running it on a directory that I've run it on before with no problems ("Files") I'm getting this error:
Error: lexical error: malformed number, a digit is required after the minus sign.
          11 12:51:47",  "CreateDate": -- press ENTER --"2024:07:11 12
                     (right here) ------^

The directory has ~25k files in several levels.

I think this is from exiftoolr, not from exiftool because I've run the exiftool command by hand and gotten no problems.

I suspect it has to do with the (-k) in the filename, but the exiftool filename seems like it may be hard-wired in exiftoolr and I can't find it.

But it has also run just fine before.

Any ideas? Any tests I can run?

The output is never saved, so I don't know how to find out which file this may be happening on (in case it's some kind of corruption)(although, as I said, I've run it successfully on this directory before).

Thanks for your time, and your time creating exiftoolr!

**Comment by jrogawa** [comment]
Also, this is the latest version of exiftoolr and exiftool running on Windows 11.

**Comment by JoshOBrien** [comment]
@jrogawa Sorry to hear that that's giving you problems. Does the call to `exif_read()` work for you when applied to other directories? 

If so, then this error likely has something to do with the names or contents of the files or directories in the directory tree to which you are applying it. In that case, I'd probably approach this by seeing if I could pinpoint which file or directory is giving you problems. So, for example, call `exif_read()` on each subdirectory of your main directory to find which of them contains the problematic content, then continue to narrow down from there.

That way, you can eventually share with me a very pared down reproducible example with the code you are running and a zipped copy of the directory and one or a few photos that cause the problem. Without that, unfortunately, I won't be able to know what the issue is and how to fix it.

**Comment by jrogawa** [comment]
Thank you so much for replying!

I’ll start testing now and let you know.

J

**Comment by jrogawa** [comment]
OK. There’s something weird about my installation. See below:

> library(exiftoolr)
>
> setwd("C:/Users/jroga/Desktop/Backup work")
>
> list.files(".")
[1] "- Start RStudio.Rproj"     "1 read data.r"
 [3] "2 fix.r"                   "3 earliest.r"
 [5] "3 latest.r"                "4 read data again.r"
 [7] "5 deal with Gs.r"          "6 read data again again.r"
[9] "7 write dates.r"           "All, original.RData"
[11] "Earliest"                  "Earliest, original.RData"
[13] "Earliest2, original.RData" "Early Tages.RData"
[15] "exiftool(-k).exe"          "exiftool.exe"
[17] "exiftool_files"            "Extras"
[19] "Files"                     "LaSals.jpg"
[21] "Latest"                    "Latest, original.RData"
[23] "Latest2, original.RData"   "Old"
[25] "tags.tab"                  "test.r"
> exif_read("LaSals.jpg")
-- press ENTER --
Using ExifTool version -- press ENTER --12.93

Error: lexical error: malformed number, a digit is required after the minus sign.
                                       -- press ENTER --[{  "SourceFil
                     (right here) ------^

So, it can’t even process “LaSals.jpg” in the same directory.

I uninstalled and reinstalled exiftoolr and then ran exiftoolr::install_exiftool() and it’s still happening.

Any ideas?

**Comment by JoshOBrien** [comment]
Unfortunately, I've got no idea what's going on there.  In particular, I don't understand why you are getting all of those "press ENTER" instructions. Do you have to actually press the Enter key to proceed beyond each of those?

Also, what do you get when you run each of the following two commands?

```r
exiftoolr::exif_version()
exiftoolr:::get_exiftool_command()
```

**Comment by jrogawa** [comment]
Yeah, it’s just weird.

Here’s what I get:

> exiftoolr::exif_version()
-- press ENTER --
Using ExifTool version -- press ENTER --12.93

[1] "-- press ENTER --" "12.93"
> exiftoolr:::get_exiftool_command()
[1] "C:\\Users\\jroga\\AppData\\Roaming/R/data/R/exiftoolr/win_exe/exiftool(-k).exe"
>

The command is an odd mix of back- and forward-slashes …

J

**Comment by jrogawa** [comment]
OK. I may have found a workaround.

I copied "exiftool(-k).exe" to "exiftool.exe" in the same Appdata\Roaming directory, and then set the environment variable ("ET_EXIFTOOL_PATH") to point to it, the non-"(-k)" version.

Now everything runs fine. I can use exif_read to read the "LaSals.jpg" file info, and my 25k recursive directory structure run works without erroring out.

Tomorrow (it's late here) I'll try to set the Perl env variable and see if it works with the "(-k)" version of the file.

Hopefully this is just some strange aspect of my computing environment.

**Comment by jrogawa** [comment]
OK. I tried it with ET_PERL_PATH set and ET_EXIFTOOL_PATH set to point to the (-k) version and I got the errors. So I went back to the non-(-k) version for ET_EXIFTOOL_PATH and it works again.

I don't know what's going on, but it works this way. :-)

Let me know if you want me to run any tests, but otherwise I'm OK.

**Comment by JoshOBrien** [comment]
@jrogawa I'm glad that you found a workaround. Also, thanks for doing all of that sleuthing and reporting the details to me.

For now, I'll hope that this is a one-off issue, having to do with your computing environment and yours alone, but if I get any other reports I may end up pinging you to ask if you'd run a test or two.

Closing for now, with fingers crossed :).

**Comment by jrogawa** [comment]
Sounds like a plan.

Yep, let me know if I can be of any help.

I do really appreciate your work on this. Having programmatic access to exiftool in R is so powerful! :-)

**Comment by ajlyons** [comment]
Sorry to reopen this, but I'm getting the same error on my Windows machine. The following sample code from the `exif_read()` help page generates an error:

```
files <- dir(system.file(package = "exiftoolr", "images"), pattern = "LaSals|Lizard", full.names = TRUE)
exif_read(files)

Error: lexical error: malformed number, a digit is required after the minus sign.
                                       -- press ENTER --[{  "SourceFil
                     (right here) ------^
```

The underlying issue is the "-- press ENTER --" text that _exiftool(-k).exe_ generates at the end of the output. I've been using exifr for years and only saw the "-- press ENTER --" message start to appear within the last year. Not sure why, but a [post](https://exiftool.org/forum/index.php?topic=8676.0) by Phil Harvey explains:

_The "-- press RETURN --" pause is a result of using the -k option.  Make sure you aren't using this option, and that you have renamed "exiftool(-k).exe" to "exiftool.exe"._

Existing Windows users can use [@jrogawa](https://github.com/jrogawa)'s solution, by manually renaming "exiftool(-k).exe" to "exiftool.exe". I didn't have to update any environment variables, `configure_exiftoolr()` found the renamed file right away. If manually renaming the _exiftool(-k).exe_ file isn't viable (e.g. because it's being used in a package), a programmatic solution could be to copy it to the temp folder.

A package update could fix the problem by renaming the executable file after copying it to the install location. That seems like a safe change. Based on the [Windows installation instructions](https://exiftool.org/install.html#Windows) from Phil, the only reason to keep the "(-k)" in the file name is if you plan to drag-n-drop files on the executable file in Windows Explorer (I think!).

**Comment by jrogawa** [comment]
Thank you Josh for keeping on with this!

I haven’t used it for a while. I had a big project, but now everything is sorted.

I can test if you need it tho.

**Comment by JoshOBrien** [comment]
@jrogawa and @ajlyons . I've now pushed up to GitHub a new version of the package in which `install_exiftool()` installs an executable named `"exiftool.exe"` instead of one named `"exiftool(-k).exe"`. It should fix the problem that both of you reported (thanks!), while not *requiring* other users who upgrade to this and future package versions to reinstall an ExifTool executable.

If one or both of you are willing to test the new version, you can install it by doing the following:

```r
devtools::install_github("JoshOBrien/exiftoolr")
```

**Comment by ajlyons** [comment]
Thank you @JoshOBrien! The package update solved the issue for me. Of course even after updating the package I had to remember to re-run `install_exiftool()` to get the the executable without the `(-k)` installed, but that worked flawlessly. 

Curious enough (but moot at this point), the errors I was getting when the -k parameter was hardwired into the executable filename were also somewhat inconsistent. When I ran `exif_read(files)` in a script, I got the error above. But in an Rmarkdown code chunk it worked fine. The `"Lexical error..."` message suggests the '-- press ENTER --' message was maybe being parsed as JSON, so maybe that was being handled different at the console vs the rendered markdown. In my own package, [uasimg](https://ucanr-igis.github.io/uasimg/), I [run exiftool](https://github.com/UCANR-IGIS/uasimg/blob/ad578a7746d1780dc532c5ea24224c5c5d0107eb/R/uas_info.R#L430C6-L430C102) using `system2` and send the output directly to a temporary CSV file, which has never failed even when the '-- press ENTER --' output could not be suppressed. 

Thanks again for publishing the update so quickly. Writing wrappers for a CLI program is hard enough, writing packages that have to interact with a user's file system is even harder, and making it work across platforms adds even more challenges. So getting it all to work is quite a feat.
